components/brave_today/browser/brave_news_p3a.cc

```diff
@@ -48,11 +50,69 @@ void RecordToHistogramBucket(const char* histogram_name,
   base::UmaHistogramExactLinear(histogram_name, answer, std::size(buckets) + 1);
 }
 
-}  // namespace
+void RecordLastUsageTime(PrefService* prefs) {
+  base::Time last_session_time =
+      prefs->GetTime(prefs::kBraveTodayLastSessionTime);
+  if (last_session_time.is_null()) {
+    return;
+  }
+  int answer = 0;
+  int duration_days = (base::Time::Now() - last_session_time).InDays();
+  int duration_weeks = duration_days / 7;
+  if (duration_weeks < 4) {
+    answer = duration_weeks + 1;
+  } else {
+    int duration_months = duration_days / 30;
+    answer = duration_months < 2 ? 5 : 6;
+  }
+  UMA_HISTOGRAM_EXACT_LINEAR(kLastUsageTimeHistogramName, answer, 7);
+}
+
+void RecordNewUserReturning(PrefService* prefs) {
+  base::Time last_use_time = prefs->GetTime(prefs::kBraveTodayLastSessionTime);
+  base::Time first_use_time =
+      prefs->GetTime(prefs::kBraveTodayFirstSessionTime);
+  int answer = 0;
+  if (!first_use_time.is_null()) {
+    bool prev_used_second_day =
+        prefs->GetBoolean(prefs::kBraveTodayUsedSecondDay);
+    int first_now_delta_days = (base::Time::Now() - first_use_time).InDays();
+    int first_last_delta_days = (last_use_time - first_use_time).InDays();
+    if (first_now_delta_days >= 7) {
+      // I used Brave News, but I'm not a first time user this week
+      answer = 1;
+    } else if (first_last_delta_days == 0) {
+      // I'm a first time user this week, but did not return again during the
+      // week
+      answer = 2;
+    } else if (prev_used_second_day || first_last_delta_days == 1) {
+      // I'm a first time user this week, and I returned the following day
+      answer = 3;
+      if (!prev_used_second_day) {
+        prefs->SetBoolean(prefs::kBraveTodayUsedSecondDay, true);
```

Comment:
Nothing ever resets this preference, so it seems like after this path is taken, the function will always record `3` instead of any higher value?

Regardless might be better to move the setting code to one of the higher-level `RecordAt` functions, or to a separate helper, like you do with `ResetCurrSessionTotalViewsCount`. Just because it's easier to reason about a function that only reads prefs, without writing them.

Reply:
It will record `3` until the end of the week. After the week ends, `1` will be recorded.

I'm not sure if I'm convinced that moving the `SetBoolean` line will improve things. If I were to write a separate helper, it would be a one-liner. I think it makes sense to keep it as is, but feel free to elaborate if I'm missing something.

Reply:
Ok, I think I see how you're intending this to work. It does seem brittle though. Please add some more comments about the behaviour you're depending on so it's clear how to fix it when it breaks. e.g.
```
// First and Last use times are recorded at session start. We're only
// interested in the first week of use, so we rely on those to report 0
// before and 1 after that first week. For the same reason, it's safe
// to leave kBraveTodayUsedSecondDay set forever.
```

Reply:
I think you could move the second-day set logic to `RecordAtSessionStart` where you're already setting the two timestamps it depends on.

Having the set calls all in one place makes it easier to reason about the data flow, which can make it easier to maintain in the future. Even with code I wrote myself, I often forget the details by the time I have to fix a bug or update behaviour after a few months. Anything that speeds up understanding the implementation pays off later, so I find it a helpful style guideline. But I don't feel strongly about it in this case.

Reply:
I can see why having all the set calls in one place would improve code visibility, but the tradeoff is that we would have to move the time delta calculations and conditionals to `RecordAtSessionStart` which may increase scope creep for that function